src/state_circuit/state.rs

```diff
@@ -0,0 +1,1609 @@
+                            || Ok(F::from_u64(idx as u64)),
+                        )?;
+                    }
+                    Ok(())
+                },
+            )
+            .ok();
+
+        layouter
+            .assign_region(
+                || "padding value table",
+                |mut region| {
+                    for idx in 0..=1 {
+                        region.assign_fixed(
+                            || "padding value table",
+                            self.padding_value_table,
+                            idx,
+                            || Ok(F::from_u64(idx as u64)),
+                        )?;
+                    }
+                    Ok(())
+                },
+            )
+            .ok();
+
+        layouter
+            .assign_region(
+                || "memory address table with zero",
+                |mut region| {
+                    for idx in 0..=MEMORY_ADDRESS_MAX {
+                        region.assign_fixed(
+                            || "address table with zero",
+                            self.memory_address_table_zero,
+                            idx,
+                            || Ok(F::from_u64(idx as u64)),
+                        )?;
+                    }
+                    Ok(())
+                },
+            )
+            .ok();
+
+        layouter.assign_region(
+            || "stack address table with zero",
+            |mut region| {
+                for idx in 0..=STACK_ADDRESS_MAX {
+                    region.assign_fixed(
+                        || "stack address table with zero",
+                        self.stack_address_table_zero,
+                        idx,
+                        || Ok(F::from_u64(idx as u64)),
+                    )?;
+                }
+                Ok(())
+            },
+        )
+    }
+
+    fn assign_operations(
+        &self,
+        ops: Vec<Op<F>>,
+        max_rows: usize,
+        target: usize,
+        address_diff_is_zero_chip: &IsZeroChip<F>,
+        start_offset: usize,
+        region: &mut Region<F>,
+    ) -> Result<Vec<BusMapping<F>>, Error> {
+        let mut ops_num = 0;
+        if target == 2 {
+            // init rows need to be counted
+            ops_num = ops
+                .clone()
+                .into_iter()
+                .fold(0, |acc, x| acc + x.global_counters.len() + 1);
+        } else if target == 3 {
+            ops_num = ops
+                .clone()
+                .into_iter()
+                .fold(0, |acc, x| acc + x.global_counters.len())
+        }
+
+        if ops_num > max_rows {
+            panic!("too many operations for the specified fixed value of operations");
+        }
+
+        let mut bus_mappings: Vec<BusMapping<F>> = Vec::new();
+
+        let mut offset = start_offset;
+        for (index, op) in ops.iter().enumerate() {
+            let address = op.address;
+            let mut address_prev = ops.first().unwrap().address;
```

Comment:
address_prev line  can move above outside "for" loop to reduce redundant assignment times according to ops.size()

Reply:
True, will change it. Tnx

Reply:
Just checked it and saw why I put it there - when ops is empty, there is no ops.first(). Could be written differently, but it's not that important I guess.